pkg/controller/pv_monitor_controller.go

```diff
@@ -0,0 +1,248 @@
+
+	ctrl := &PVMonitorController{
+		csiConn:            conn,
+		eventRecorder:      eventRecorder,
+		supportListVolumes: supportListVolumes,
+		enableNodeWatcher:  enableNodeWatcher,
+		client:             client,
+		monitorName:        monitorName,
+		monitorInterval:    monitorInterval,
+		pvQueue:            workqueue.NewNamed("csi-monitor-pv-queue"),
+		//pvQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "csi-monitor-pv-queue"),
+
+		pvcToPodsCache: util.NewPVCToPodsCache(),
+		pvEnqueued:     make(map[string]bool),
+	}
+
+	// PV informer
+	pvInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ctrl.pvAdded,
+		// we do not care about PV changes, so do not need UpdateFunc here.
+		// deleted PVs will not be readded to the queue, so do not need DeleteFunc here
+	})
+	ctrl.pvLister = pvInformer.Lister()
+	ctrl.pvListerSynced = pvInformer.Informer().HasSynced
+
+	// PVC informer
+	ctrl.pvcLister = pvcInformer.Lister()
+	ctrl.pvListerSynced = pvcInformer.Informer().HasSynced
+
+	// Pod informer
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: ctrl.podAdded,
+		// UpdateFunc: ctrl.podUpdated,  TODO: do we need this ?
+		DeleteFunc: ctrl.podDeleted,
+	})
+	ctrl.podLister = podInformer.Lister()
+	ctrl.podListerSynced = podInformer.Informer().HasSynced
+
+	if enableNodeWatcher {
+		ctrl.nodeWatcher = NewNodeWatcher(ctrl.monitorName, ctrl.client, ctrl.pvLister, ctrl.pvcLister, nodeInformer, ctrl.eventRecorder, ctrl.pvcToPodsCache, monitorInterval)
+	}
+
+	ctrl.pvChecker = handler.NewPVHealthConditionChecker(monitorName, conn, client, timeout, ctrl.pvcLister, ctrl.pvLister, ctrl.eventRecorder)
+
+	return ctrl
+}
+
+// Run runs the volume health condition checking method
+func (ctrl *PVMonitorController) Run(workers int, stopCh <-chan struct{}) {
+	defer ctrl.pvQueue.ShutDown()
+
+	klog.Infof("Starting CSI External PV Health Monitor Controller")
+	defer klog.Infof("Shutting down CSI External PV Health Monitor Controller")
+
+	if !cache.WaitForCacheSync(stopCh, ctrl.pvcListerSynced, ctrl.pvListerSynced, ctrl.podListerSynced) {
+		klog.Errorf("Cannot sync cache")
+		return
+	}
+
+	if ctrl.enableNodeWatcher {
+		go ctrl.nodeWatcher.Run(stopCh)
+	}
+
+	if ctrl.supportListVolumes {
+		go wait.Until(ctrl.checkPVsHealthConditionByListVolumes, listVolumesInterval, stopCh)
+	} else {
+		for i := 0; i < workers; i++ {
+			go wait.Until(ctrl.checkPVWorker, ctrl.monitorInterval, stopCh)
+		}
+
+		go wait.Until(func() {
+			err := ctrl.ReconcilePVs()
+			if err != nil {
+				klog.Errorf("Failed to reconcile volumes: %v", err)
+			}
+		}, pvReconcileSyncInterval, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (ctrl *PVMonitorController) checkPVsHealthConditionByListVolumes() {
+	err := ctrl.pvChecker.CheckControllerVolumesStatus()
+	if err != nil {
+		klog.Errorf("check controller volume status error: %+v", err)
+	}
+}
+
+// ReconcilePVs reconciles PVs
```

Comment:
Reconcile means it keeps trying to bring current state to desired state, but this function mainly add PVs to the queue.  It does not really process the PV to bring it to desired state.  So I think this function should be renamed to something else, i.e., AddPVsToQueue.  Also please add comments.